src/timesync/timesyncd.c

```diff
@@ -70,7 +70,7 @@ static int load_clock_timestamp(uid_t uid, gid_t gid) {
 
         } else {
                 r = mkdir_safe_label("/var/lib/systemd/timesync", 0755, uid, gid, true);
-                if (r < 0)
+                if (r < 0 && r != -EEXIST)
```

Comment:
uh, no, this is not right.

mkdir_safe() return EEXIST only if the dir exists and doesn't have compatible access mode/ownership. If it already exists and the access mode/ownership are good, then it will return 0 cleanly. It appears to me that this patch is hence unnecessary, and tapes over security problems.

Reply:
This is the same as the first version I've implemented in #7003. See the discussion [here](https://github.com/systemd/systemd/pull/7003#discussion_r142891599). I've introduced a flag to follow symlinks in `mkdir_safe_label()` in the PR. So, if `mkdir_safe_label()` succeeds, now `/var/lib/systemd/timesync` is a directory or a link to a directory with the expected mode, uid and gid.

Reply:
After looking at @poettering's and @yuwata's comments, I think my patch is good ;)
Basically, the admin may for her own reasons create this directory with a higher a lower access mode. Also, as we can see in the PRs, this directory might have been created by us or packaging scripts or whatnot. There only thing that matters is whether we can create the state file or not, which we try to do in the next line. This is the pattern we generally follow when creating directories, and I don't see a reason to deviate here.

Reply:
Still I do not know the detail of `boot-and-services` test. But if the test runs `systemd-timesyncd` with the provided .service file, then the directory is created with or, if it already exists, set correct mode, uid and gid by pid1. So, I do not understand why timesyncd fails here...

Reply:
I can perfectly reproduce this in a VM without any test. Manually cleaning out the state dir and starting the service reproduces it already - apparently the act of starting the service already creates it twice. Presumably once it's from systemd itself from `StateDirectory=`, and another time from timesyncd in this code. So systemd is stepping on its own toes here?

Reply:
@martinpitt but it appears we should just set StateDirectoryMode= correctly then? normally, PID 1 and the service itself should really set up the directory the same way and if we don#t something is really off

Reply:
@poettering: Absolutely; I'm not defending this particular implementation, I was just curious whether this fixes the tests, and wanted to point out that this is not something test specific - the tests don't do any magic, they just boot up the system and verify that there are no failed services. In other words, this is an actual bug.